libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+
+/// Allows one to control the behaviour of the floodsub system.
+#[derive(Clone)]
+pub struct FloodSubController<T, C>
+where
+    T: MuxedTransport + 'static,                // TODO: 'static :-/
+    C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :-/
+{
+    inner: Arc<Inner>,
+    swarm: SwarmController<T, C>,
+}
+
+struct Inner {
+    // Our local peer ID multihash, to pass as the source.
+    peer_id: Vec<u8>,
+
+    // Channel where to send the messages that should be dispatched to the user.
+    output_tx: mpsc::UnboundedSender<Message>,
+
+    // Active connections with a remote.
+    remote_connections: RwLock<FnvHashMap<Multiaddr, RemoteInfo>>,
+
+    // List of topics we're subscribed to. Necessary in order to filter out messages that we
+    // erroneously receive.
+    subscribed_topics: RwLock<Vec<Topic>>,
+
+    // Sequence number for the messages we send.
+    seq_no: AtomicUsize,
+
+    // We keep track of the messages we received (in the format `(remote ID, seq_no)`) so that we
+    // don't dispatch the same message twice if we receive it twice on the network.
+    // TODO: the `HashSet` will keep growing indefinitely :-/
+    received: Mutex<FnvHashSet<(Vec<u8>, Vec<u8>)>>,
+}
+
+struct RemoteInfo {
+    // Sender to send data over the socket to that host.
+    sender: mpsc::UnboundedSender<BytesMut>,
+    // Topics the remote is registered to.
+    subscribed_topics: RwLock<FnvHashSet<TopicHash>>,
+}
+
+impl fmt::Debug for Inner {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Inner")
+            .field("peer_id", &self.peer_id)
+            .field(
+                "num_remote_connections",
+                &self.remote_connections.read().len(),
+            )
+            .field(
+                "num_subscribed_topics",
+                &self.subscribed_topics.read().len(),
+            )
+            .field("seq_no", &self.seq_no)
+            .field("received", &self.received)
+            .finish()
+    }
+}
+
+impl<T, C> FloodSubController<T, C>
+where
+    T: MuxedTransport + 'static,                // TODO: 'static :-/
+    C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :-/
+{
+    /// Builds a new controller for floodsub.
+    #[inline]
+    pub fn new(
+        upgrade: &FloodSubUpgrade,
+        swarm: SwarmController<T, C>,
+    ) -> FloodSubController<T, C> {
+        FloodSubController {
+            inner: upgrade.inner.clone(),
+            swarm: swarm,
+        }
+    }
+
+    /// Subscribe to a topic. When a node on the network sends a message for that topic, we will
+    /// likely receive it.
+    ///
+    /// It is not guaranteed that we receive every single message published on the network.
+    #[inline]
+    pub fn subscribe(&self, topic: &Topic) {
+        // This function exists for convenience.
+        self.subscribe_multi(iter::once(topic));
+    }
+
+    /// Same as `subscribe`, but subscribes to multiple topics at once.
```

Comment:
Maybe add a doc saying that `subscribe_multi`-ing to many topics is preferable to subscribing to each independently because the broadcasts get batched. Also, maybe I'm wrong but I think the standard naming of this would be `publish_many`.

I think the best way would be to add a `IntoTopics` trait and implement it for `T: IntoIterator<Item=&'a Topic>` and `&'a Topic` and then just have one method, but I would understand if you'd rather not.